syntax = "proto3";

package distributedstore;

service KeyValueStore {
  rpc put(PutRequest) returns (PutResponse);

  rpc get(GetRequest) returns (GetResponse);

  rpc slowDown(SlowDownRequest) returns (SlowDownResponse);

  rpc restore(RestoreRequest) returns (RestoreResponse);

  rpc canCommit(CanCommitRequest) returns (CommitResponse);

  rpc doCommit(DoCommitRequest) returns (CommitResponse);

  rpc doAbort(DoAbortRequest) returns (DoAbortResponse);

}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {
  bool success = 1;
}

message GetRequest {
  string key = 1;
}

message GetResponse {
  string value = 1;
  bool found = 2;
}

message SlowDownRequest {
  int32 seconds = 1;
}

message SlowDownResponse {
  bool success = 1;
}

message RestoreRequest {}

message RestoreResponse {
  bool success = 1;
}

message CanCommitRequest {
  string key = 1;
  string value = 2;
}

message DoCommitRequest {
  string key = 1;
  string value = 2;
}

message CommitResponse {
  bool success = 1;
}

message DoAbortRequest {

}

message DoAbortResponse {
  bool success = 1;
}

message Empty {

}